pallets/sto/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+        /// Total fundraisers created for a token
+        FundraiserCount get(fn fundraiser_count): map hasher(twox_64_concat) Ticker => u64;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Create a new offering. A fixed amount of pre-minted tokens are put up for sale at the specified tiered rate.
+        #[weight = 800_000_000]
+        pub fn create_fundraiser(
+            origin,
+            offering_portfolio: PortfolioId,
+            offering_asset: Ticker,
+            raising_portfolio: PortfolioId,
+            raising_asset: Ticker,
+            tiers: Vec<PriceTier<T::Balance>>,
+            venue_id: u64,
+            start: Option<T::Moment>,
+            end: Option<T::Moment>,
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(T::Asset::primary_issuance_agent(&offering_asset) == did, Error::<T>::Unauthorized);
+
+            let venue = VenueInfo::get(venue_id).ok_or(Error::<T>::InvalidVenue)?;
+            ensure!(
+                venue.creator == did && venue.venue_type == VenueType::Sto,
+                Error::<T>::InvalidVenue
+            );
+
+            <Portfolio<T>>::ensure_portfolio_custody(raising_portfolio, did)?;
+            ensure!(<PortfolioAssetBalances<T>>::contains_key(raising_portfolio, raising_asset), Error::<T>::InvalidPortfolio);
+
+            <Portfolio<T>>::ensure_portfolio_custody(offering_portfolio, did)?;
+            ensure!(<PortfolioAssetBalances<T>>::contains_key(offering_portfolio, offering_asset), Error::<T>::InvalidPortfolio);
+
+            ensure!(
+                tiers.len() > 0 && tiers.iter().all(|t| t.total > 0.into()),
+                Error::<T>::InvalidPriceTiers
+            );
+
+            let offering_amount: T::Balance = tiers
+                .iter()
+                .map(|t| t.total)
+                .fold(0.into(), |total, x| total + x);
+
+            <Portfolio<T>>::lock_tokens(&offering_portfolio, &offering_asset, &offering_amount)?;
+
+            let fundraiser_id = Self::fundraiser_count(offering_asset) + 1;
+            let fundraiser = Fundraiser {
+                    offering_portfolio,
+                    offering_asset,
+                    raising_portfolio,
+                    raising_asset,
+                    tiers: tiers.into_iter().map(Into::into).collect(),
+                    venue_id,
+                    start: start.unwrap_or_else(Timestamp::<T>::get),
+                    end,
+                    frozen: false,
+            };
+
+            if let Some(end) = fundraiser.end {
+                ensure!(fundraiser.start < end, Error::<T>::InvalidOfferingWindow);
+            }
```

Comment:
Let's do this before constructing the `Fundraiser`, to avoid unnecessarily reallocating for `tiers`.